File: prompts.py
"""LLM prompts for each stage of the CV screening pipeline.

Design principle: every prompt forces the model to act like a senior hiring
manager + recruiter + CV coach, reason semantically, and return strict JSON.
"""

JD_EXTRACTION_PROMPT = """You are a Senior Technical Recruiter with 10+ years of experience.
Read this Job Description and extract a structured understanding of it.

Return STRICT JSON only, matching this schema:
{
  "role_title": "string",
  "seniority_level": "Intern | Junior | Mid | Senior | Lead | Principal | Manager | Director",
  "domain": "string (industry / product area)",
  "required_skills": ["..."],
  "nice_to_have_skills": ["..."],
  "key_responsibilities": ["..."],
  "required_years_experience": "string (e.g. '3-5 years' or 'not specified')",
  "soft_skills": ["..."],
  "hidden_expectations": ["things recruiters usually infer but aren't explicit"]
}

Rules:
- Be specific. Do NOT invent requirements.
- Split "required" vs "nice-to-have" based on language cues ("must", "required" vs "bonus", "plus").
- "hidden_expectations" is your senior recruiter read-between-the-lines insight.

JOB DESCRIPTION:
---
{jd}
---
"""


CV_EXTRACTION_PROMPT = """You are a Senior Recruiter reading a candidate CV.
Extract a structured profile. Return STRICT JSON only:

{
  "candidate_name": "string or 'unknown'",
  "headline": "one-line profile",
  "total_years_experience": "string",
  "current_role": "string",
  "career_trajectory": "string (describe progression: ascending, lateral, scattered, etc.)",
  "domain_expertise": ["..."],
  "core_strengths": ["..."],
  "technical_skills": ["..."],
  "soft_skills": ["..."],
  "notable_achievements": ["quantified wins if any"],
  "education": ["..."],
  "bullets": [
     {"role": "string", "company": "string", "text": "original bullet text"}
  ],
  "red_flag_signals": ["short job tenures, gaps, buzzword spam, vague claims"]
}

Rules:
- Preserve original bullet text verbatim (we'll critique it later).
- Infer trajectory from dates + titles.
- Do not fabricate.

CV TEXT:
---
{cv}
---
"""


EVALUATION_PROMPT = """You are a panel of THREE senior experts reviewing a candidate for a specific role:
  1. Senior Hiring Manager (10+ yrs) — decides hire / no hire
  2. Senior Technical Recruiter — checks JD fit, ATS, positioning
  3. Senior CV Coach — rewrites weak content with impact + metrics

You already have structured JD and CV understandings below. Now produce the FINAL
evaluation as STRICT JSON. No prose outside JSON.

Scoring rubric (sum to overall_score out of 100):
  - jd_match          (0-40)  semantic + transferable skill match, not just keywords
  - cv_quality        (0-25)  bullets with impact, clarity, measurable results
  - experience_depth  (0-10)  depth vs. required seniority
  - formatting        (0-15)  structure, scannability, ATS-friendliness
  - risk              (0-10)  inverse of red flags (10 = no risk)

Grade mapping:
  85+ Strong Hire | 70-84 Good Fit | 50-69 Moderate | <50 Weak

OUTPUT SCHEMA (fill every field, arrays may be empty but must exist):
{{
  "overall_score": 0,
  "grade": "Strong Hire | Good Fit | Moderate | Weak",
  "confidence": 0.0,
  "summary": "3-5 sentence hiring-manager style verdict",
  "dimension_scores": {{
    "jd_match": 0,
    "cv_quality": 0,
    "experience_depth": 0,
    "formatting": 0,
    "risk": 0
  }},
  "strengths": ["specific, evidence-backed"],
  "weaknesses": ["specific, evidence-backed"],
  "improvements": {{
    "content_issues": [
      {{
        "issue_type": "weak_bullet | no_metrics | vague_claim | buzzword_spam",
        "original": "verbatim bullet from CV",
        "problem": "why this is weak",
        "improved_version": "rewritten bullet with impact + metrics (may use realistic placeholder numbers in [brackets])"
      }}
    ],
    "skill_gaps": {{
      "critical_missing": ["skills required by JD the candidate clearly lacks"],
      "secondary_missing": ["nice-to-haves missing"],
      "transferable": ["skills candidate has that partially cover gaps"]
    }},
    "positioning_issues": [
      {{
        "problem": "Your CV is positioned as X but role requires Y",
        "rewritten_summary": "new 2-3 sentence professional summary tailored to JD"
      }}
    ],
    "experience_issues": ["unclear progression, thin depth in X, etc."],
    "formatting_issues": ["ATS-unfriendly, hard to scan, etc."],
    "red_flags": [
      {{"flag": "string", "risk_explanation": "string"}}
    ]
  }},
  "suggestions": {{
    "micro_fixes": ["bullet-level rewrites, wording, metrics — be concrete"],
    "macro_fixes": ["section reorder, new summary, highlight X experience"],
    "strategic_advice": ["e.g. 'You lack cloud ops; get AWS cert to pass HR screens'"]
  }},
  "hiring_decision": {{
    "recommendation": "Hire | Consider | Reject",
    "reason": "crisp justification grounded in the scores and evidence",
    "top_risks": ["..."]
  }}
}}

CRITICAL RULES:
- NEVER be generic. Every weakness must cite a bullet, a missing skill, or a concrete gap.
- Every improved_version MUST be materially better: add impact verb + scope + metric.
- Semantic matching: "built REST APIs in Flask" ≈ "backend services in Python" — credit it.
- If the candidate is clearly wrong for the role, say so. Do not soften.
- dimension_scores MUST sum to overall_score.

---
JD UNDERSTANDING (JSON):
{jd_json}

---
CV UNDERSTANDING (JSON):
{cv_json}

---
RAW CV TEXT (for bullet-level quoting):
{cv_raw}
"""

File: lib/prompts.ts
// lib/prompts.ts

export const SYSTEM_PROMPT =
  "You are a panel of senior hiring experts: a Senior Hiring Manager (10+ yrs), " +
  "a Senior Technical Recruiter, a Senior CV Coach, and a Senior Career Strategist. " +
  "You reason semantically like a real recruiter, never do keyword matching, " +
  "and you ALWAYS return strict valid JSON only — no prose, no markdown fences.";

export const JD_EXTRACTION_PROMPT = (jd: string) => `
You are a Senior Technical Recruiter with 10+ years of experience.
Read this Job Description and extract a structured understanding of it.

Return STRICT JSON only, matching this schema:
{
  "role_title": "string",
  "seniority_level": "Intern | Junior | Mid | Senior | Lead | Principal | Manager | Director",
  "domain": "string",
  "required_skills": ["..."],
  "nice_to_have_skills": ["..."],
  "key_responsibilities": ["..."],
  "required_years_experience": "string",
  "soft_skills": ["..."],
  "hidden_expectations": ["..."]
}

Rules:
- Be specific. Do NOT invent requirements.
- Split required vs nice-to-have based on language cues.
- "hidden_expectations" is your senior recruiter read-between-the-lines insight.

JOB DESCRIPTION:
---
${jd}
---
`.trim();

export const CV_EXTRACTION_PROMPT = (cv: string) => `
You are a Senior Recruiter reading a candidate CV.
Extract a structured profile. Return STRICT JSON only:

{
  "candidate_name": "string or 'unknown'",
  "headline": "one-line profile",
  "total_years_experience": "string",
  "current_role": "string",
  "career_trajectory": "string",
  "domain_expertise": ["..."],
  "core_strengths": ["..."],
  "technical_skills": ["..."],
  "soft_skills": ["..."],
  "notable_achievements": ["..."],
  "education": ["..."],
  "bullets": [{"role": "string", "company": "string", "text": "string"}],
  "red_flag_signals": ["..."]
}

CV TEXT:
---
${cv}
---
`.trim();

export const EVALUATION_PROMPT = (
  jdJson: string,
  cvJson: string,
  cvRaw: string,
  position: string,
  level: string,
  compareMarket: boolean
) => {
  const marketSchemaBlock = compareMarket
    ? `,
  "market_comparison": {
    "market_rank_top_pct": 0,
    "market_context": "2-3 sentence analysis of how this candidate compares to the current ${position || "this role"} market at ${level || "this"} level. Be specific about what makes them above or below average.",
    "market_trends": [
      "List 4-6 concrete things that top companies currently look for in a ${position || "this role"} at ${level || "this"} level — beyond just this JD. Think broadly: architecture patterns, methodologies, tools, soft skills, certifications, open-source contributions, etc."
    ],
    "skill_alignment": [
      "List 3-5 concrete, actionable steps this candidate should take to become 'what the market needs'. Each step must reference the candidate's current profile and explain the gap."
    ],
    "improvement_priority": [
      "Top 3 highest-impact changes this candidate should make FIRST to maximize their market value. Be specific and actionable."
    ],
    "smart_action_plan": [
      {"phase": "Next 30 days", "actions": ["2-3 immediate, concrete actions"]},
      {"phase": "Next 3 months", "actions": ["2-3 medium-term skill-building actions"]},
      {"phase": "Next 6-12 months", "actions": ["2-3 long-term career positioning actions"]}
    ]
  }`
    : "";

  const marketRulesBlock = compareMarket
    ? `
- MARKET COMPARISON RULES (CRITICAL):
  - "market_rank_top_pct" is an integer 1-100. A value of 15 means "top 15%" of candidates.
  - Base your ranking on the CURRENT job market for "${position || "this role"}" at "${level || "this"}" level.
  - "market_trends" must go BEYOND the provided JD. Think about what Google, Meta, top startups, and leading companies in this domain currently seek. Include emerging technologies, architectural patterns, methodologies.
  - "skill_alignment" must be personalized to THIS candidate: reference their actual skills and explain precisely what's missing vs what the market wants.
  - "smart_action_plan" must be a realistic, phased roadmap. Each action must be concrete (e.g., "Complete AWS Solutions Architect certification" not "Learn cloud").
  - DO NOT be generic. Every item must be tailored to THIS candidate's profile and THIS market.`
    : "";

  return `
You are a panel of ${compareMarket ? "FOUR" : "THREE"} senior experts reviewing a candidate for a specific role:
  1. Senior Hiring Manager (10+ yrs) — decides hire / no hire
  2. Senior Technical Recruiter — checks JD fit, ATS, positioning
  3. Senior CV Coach — rewrites weak content with impact + metrics${
    compareMarket
      ? "\n  4. Senior Career Strategist — analyzes broader market trends, competitive positioning, and builds actionable growth roadmaps"
      : ""
  }

ROLE CONTEXT:
- Targeted Position: ${position || "Not specified"}
- Seniority Level: ${level || "Not specified"}

Scoring rubric (sum to overall_score out of 100):
  - jd_match (0-40), cv_quality (0-25), experience_depth (0-10), formatting (0-15), risk (0-10)

Grade mapping: 85+ Strong Hire | 70-84 Good Fit | 50-69 Moderate | <50 Weak

OUTPUT SCHEMA (fill every field):
{
  "overall_score": 0,
  "grade": "Strong Hire | Good Fit | Moderate | Weak",
  "confidence": 0.0,
  "summary": "3-5 sentence verdict",
  "dimension_scores": {"jd_match":0,"cv_quality":0,"experience_depth":0,"formatting":0,"risk":0},
  "strengths": ["..."],
  "weaknesses": ["..."],
  "improvements": {
    "content_issues": [{"issue_type":"weak_bullet|no_metrics|vague_claim|buzzword_spam","original":"...","problem":"...","improved_version":"..."}],
    "skill_gaps": {"critical_missing":[],"secondary_missing":[],"transferable":[]},
    "positioning_issues": [{"problem":"...","rewritten_summary":"..."}],
    "experience_issues": ["..."],
    "formatting_issues": ["..."],
    "red_flags": [{"flag":"...","risk_explanation":"..."}]
  },
  "suggestions": {"micro_fixes":[],"macro_fixes":[],"strategic_advice":[]},
  "hiring_decision": {"recommendation":"Hire|Consider|Reject","reason":"...","top_risks":[]}${marketSchemaBlock}
}

CRITICAL RULES:
- NEVER be generic. Every weakness must cite a bullet or concrete gap.
- dimension_scores MUST sum to overall_score.${marketRulesBlock}

JD UNDERSTANDING: ${jdJson}
CV UNDERSTANDING: ${cvJson}
RAW CV TEXT: ${cvRaw.slice(0, 8000)}
`.trim();
};
